HWLesson 4.6 (Module)/MigrationProject/Services/LocationService.cs

```diff
@@ -0,0 +1,130 @@
+﻿using Data;
+using Microsoft.Extensions.Logging;
+using Models;
+using Repositories.Abstractions;
+using Services.Abstractions;
+
+namespace Services
+{
+    public class LocationService : BaseDataService<AppDbContext>, ILocationService
+    {
+        private readonly ILocationRepository _locationRepository;
+        private readonly ILogger<LocationService> _loggerService;
+
+        public LocationService(IDbContextWrapper<AppDbContext> dbContextWrapper,
+             ILogger<BaseDataService<AppDbContext>> logger,
+             ILocationRepository locationRepository,
+             ILogger<LocationService> loggerService)
+            : base(dbContextWrapper, logger)
+        {
+            _locationRepository = locationRepository;
+            _loggerService = loggerService;
+        }
+
+        public async Task<int> AddLocationAsync(string locationName)
+        {
+            var location = await _locationRepository.GetLocationAsync(locationName);
+
+            if (location != null)
+            {
+                _loggerService.LogWarning($"Location with a name:{locationName} exists");
+                return 0;
+            }
+
+            var id = await _locationRepository.AddLocationAsync(locationName);
+            _loggerService.LogInformation($"Created Location with id = {id}");
+            return id;
+        }
+
+        public async Task<Location> GetLocationAsync(string locationName)
+        {
+            var location = await _locationRepository.GetLocationAsync(locationName);
+
+            if (location == null)
+            {
+                _loggerService.LogWarning($"Location with a name:{locationName} not founded");
+                return null!;
+            }
+
+            return new Location()
+            {
+                Id = location.Id,
+                LocationName = location.LocationName
+            };
+        }
+
+        public async Task<Location> GetLocationAsync(int id)
+        {
+            var location = await _locationRepository.GetLocationAsync(id);
+
+            if (location == null)
+            {
+                _loggerService.LogWarning($"Location with ID:{id} not founded");
+                return null!;
+            }
+
+            return new Location()
+            {
+                Id = location.Id,
+                LocationName = location.LocationName
+            };
+        }
```

Comment:
There are three almost similar methods, maybe it will be better  to combine them and add logic to create Func using reflection and pass this Func to the method in Repo